feat(snap): add startup-duration and startup-interval configure options

siggiskulason opened the pull request:
To change the default startup duration use:
sudo snap set edgex-device-camera startup-duration=n

To change the default startup interval use:
sudo snap set edgex-device-camera startup-interval=n

Signed-off-by: Siggi Skulason <siggi.skulason@canonical.com>

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://wiki.edgexfoundry.org/display/FA/Contributor%27s+Guide
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?
<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [x] Documentation content changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior and link to a relevant issue. -->
The module go-mod-bootstrap implements a service dependency check on startup which ensures that all of the runtime dependencies of a particular service are met before the service transitions to active state.

The logic implements a basic retry loop with a maximum duration and sleep interval. If the dependent services are not available, the service sleeps for the defined interval (default: 1s) and then tries again up to a maximum duration (default: 60s). These values can be overridden via environment variables. 

Issue Number:
N/A

## What is the new behavior?
This PR covers updating all the device service snap to allow these variables to be set/queried via the configure hook.

To change the default startup duration use:
sudo snap set edgex-device-camera startup-duration=n

To change the default startup interval use:
sudo snap set edgex-device-camera startup-interval=n
## Does this PR introduce a breaking change?
<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->

- [ ] Yes
- [x] No

## Are there any new imports or modules? If so, what are they used for and why?
N/A

## Are there any specific instructions or things that should be known prior to reviewing?
N/A

## Other information
N/A

lenny-intel reviewed (dismissed):
LGTM, merge per @tonyespy approval

tonyespy reviewed (changes requested):
Looks good, however there are two inline comments that need to be addressed. Also your branch is out-of-date with master, so you need to update it.

Regarding your PR description, if there are questions (e.g. Are there any specific instructions..., Other information) in the PR template that don't apply to your changes, then I either add **NA**, or delete the question. You should always answer the question about new imports though, again in this case with **NA**.

I tested both new config options and they work as expected. I see the log messages that the overrides have been applied, and both settings seem to work as expected.

That said, what I wasn't able to do was start device-camera (duration=120 interval=10) before it's dependent services (e.g. consul, core-data, core-metadata) and then start its dependent services before duration expired, resulting in a happy service. When I tried this (using snap stop/start edgexfoundry), the services never seemed to come up healthy. I tried a narrower test case where I just disabled core-metadata and brought it back up and that worked fine. So this may be related to stopping and starting all services in the snap rather than an issue with the bootstrap dependency logic. I'd be interested to see if you see any similar behavior while testing. If you do, we should file an issue so that we can follow-up on it.

With all that said, I don't think the behavior I observed should block this PR from landing.

tonyespy commented:
Also super nit-picky, but I'd re-phrase your commit message as you're not adding new configure **hooks**, you're adding new configure **options**.

siggiskulason commented on snap/README.md:
````diff
@@ -0,0 +1,30 @@
+
+The service can then be started as follows:
+```bash
+$ sudo snap start edgex-device-camera.device-camera-go
````
Changed to:  "The service can then be started as follows. The "--enable" option ensures that as well as starting the service now, it will be automatically started on boot:"

siggiskulason commented on snap/README.md:
```diff
@@ -0,0 +1,30 @@
+
+Device services implement a service dependency check on startup which ensures that all of the runtime dependencies of a particular service are met before the service transitions to active state.
+
+The logic implements a basic retry loop with a maximum duration and sleep interval. If the dependent services are not available, the service sleeps for the defined interval (default: 1s) and then tries again up to a maximum duration (default: 60s). These values can be overridden with the following commands:
```
Added:
Snapd doesn't support orchestration between services in different snaps. It is therefore possible on a reboot for a device service to come up faster than all of the required services running in the main edgexfoundry snap. If this happens, it's possible that the device service repeatedly fails startup, and if it exceeds the systemd default limits, then it might be left in a failed state. This situation might be more likely on constrained hardware (e.g. RPi).

siggiskulason commented:
> Also super nit-picky, but I'd re-phrase your commit message as you're not adding new configure **hooks**, you're adding new configure **options**.

Done - in the commit message and PR title

siggiskulason commented:
Thank you for the review. I have done the following:
- [x] rebased
- [x] updated comment with NA where relevant
- [x] README line 13: Added info about why we are doing this
- [x] Readme line 29: Added the --enable flag and information about it
- [x] re-phrase commit message as I'm adding new configure options.

siggiskulason commented:
> That said, what I wasn't able to do was start device-camera (duration=120 interval=10) before it's dependent services (e.g. consul, core-data, core-metadata) and then start its dependent services before duration expired, resulting in a happy service. When I tried this (using snap stop/start edgexfoundry), the services never seemed to come up healthy. I tried a narrower test case where I just disabled core-metadata and brought it back up and that worked fine. So this may be related to stopping and starting all services in the snap rather than an issue with the bootstrap dependency logic. I'd be interested to see if you see any similar behavior while testing. If you do, we should file an issue so that we can follow-up on it.

I was not able to reproduce this issue. I set up the snap with those values (120/10) and tried both just stopping core-metadata and also removing the edgexfoundry snap and reinstalling it. In both cases I got to a point where there are a few " edgex-core-metadata service not healthy..." messages, followed by "Check Metadata service's status via Registry..." and then success with "Service clients initialize successful." and "Camera device service started".

tonyespy reviewed (approved):
Thanks for the changes. LGTM
